packages/core/src/modules/dcql/DcqlService.ts

```diff
@@ -0,0 +1,277 @@
+   */
+  public selectCredentialsForRequest(dcqlQueryResult: DcqlQueryResult): DcqlCredentialsForRequest {
+    if (!dcqlQueryResult.canBeSatisfied) {
+      throw new DcqlError(
+        'Cannot select the credentials for the dcql query presentation if the request cannot be satisfied'
+      )
+    }
+
+    const credentials: DcqlCredentialsForRequest = {}
+
+    if (dcqlQueryResult.credential_sets) {
+      for (const credentialSet of dcqlQueryResult.credential_sets) {
+        // undefined defaults to true
+        if (credentialSet.required === false) continue
+        const firstFullFillableOption = credentialSet.options.find((option) =>
+          option.every((credential_id) => dcqlQueryResult.credential_matches[credential_id].success)
+        )
+
+        if (!firstFullFillableOption) {
+          throw new DcqlError('Invalid dcql query result. No option is fullfillable')
+        }
+
+        for (const credentialQueryId of firstFullFillableOption) {
+          const credential = dcqlQueryResult.credential_matches[credentialQueryId]
+
+          if (credential.success && credential.record.type === 'MdocRecord' && 'namespaces' in credential.output) {
+            credentials[credentialQueryId] = {
+              credentialRecord: credential.record,
+              disclosedPayload: credential.output.namespaces,
+            }
+          } else if (credential.success && credential.record.type !== 'MdocRecord' && 'claims' in credential.output) {
+            credentials[credentialQueryId] = {
+              credentialRecord: credential.record,
+              disclosedPayload: credential.output.claims,
+            }
+          } else {
+            throw new DcqlError('Invalid dcql query result. Cannot auto-select credentials')
+          }
+        }
+      }
+    } else {
+      for (const credentialQuery of dcqlQueryResult.credentials) {
+        const credential = dcqlQueryResult.credential_matches[credentialQuery.id]
+        if (credential.success && credential.record.type === 'MdocRecord' && 'namespaces' in credential.output) {
+          credentials[credentialQuery.id] = {
+            credentialRecord: credential.record,
+            disclosedPayload: credential.output.namespaces,
+          }
+        } else if (credential.success && credential.record.type !== 'MdocRecord' && 'claims' in credential.output) {
+          credentials[credentialQuery.id] = {
+            credentialRecord: credential.record,
+            disclosedPayload: credential.output.claims,
+          }
+        } else {
+          throw new DcqlError('Invalid dcql query result. Cannot auto-select credentials')
+        }
+      }
+    }
+
+    return credentials
+  }
+
+  public validateDcqlQuery(dcqlQuery: DcqlQuery.Input | DcqlQuery) {
+    return DcqlQuery.parse(dcqlQuery)
+  }
+
+  // TODO: this IS WRONG
+  private createPresentationFrame(obj: Record<string, JsonValue>): IPresentationFrame {
```

Comment:
Is this used?

Reply:
Yes for SdJwt presentations. Will fix it in the weekend